doc/architectural/goto-program-transformations.md

```diff
@@ -0,0 +1,311 @@
+
+
+
+\section optional-transforms Optional Transformation Passes
+
+The sections lists the optional transformation passes that are optional and will
+modify a goto model. Note that these are documented here for consistency, but
+not required for core goto format.
+
+Note for each optional pass there is a listed predeceesor pass. This is the pass
+currently called before the listed pass in CBMC. While the ordering may not be
+fully documented, this should be followed.
+
+\subsection string-instrument-transform String Instrumentation
+
+This (optional) pass adds checks on calls to some of the C standard library
+string functions. It uses the tracking information added by the
+\ref string-abstraction-transform pass. It is switched on by the
+`--string-abstraction` command line option.
+
+The implementation of this pass is called via the
+\ref string_instrumentation(goto_modelt &goto_model) function. Detailed
+documentation of this pass belongs in \ref string_instrumentation.h
+
+<em>The predecessor pass is \ref linking-transform .</em>
+
+\subsection inlining-transform Partial Inlining
+
+This pass does partial inlining when this option is switched on. Partal inlining
+is inlining of functions either: marked as inline, or smaller than a specified
+limit. For further detail see the implementation function \ref
+goto_partial_inline(goto_modelt &, message_handlert &, unsigned, bool)
+
+<em>Predecessor pass is \ref precondition-transform.</em>
+
+
+\subsection assertions-transform Transform Assertions Assumptions
+ 
+This pass removes user provided assertions and assumptions when user has opted
+to do so. Note that this pass removes skip instructions if any other changes
+are made. The implementation of this pass is called via the \ref
+transform_assertions_assumptions(const optionst &, goto_modelt &) function.
+
+<em>Predecessor pass is \ref check-c-transform.</em>
+
+
+\subsection string-abstraction-transform String Abstraction
+
+This (optional) transformation pass adds tracking of length of C strings. It is
+switched on by the `--string-abstraction` command line option. This changes made
+by this pass are documented as part of the documentation for the \ref
+string_abstractiont class. The implementation of this pass is called via the
+\ref string_abstraction(goto_modelt &, message_handlert &) function.
+
+<em>Predecessor pass is \ref floats-transform.</em>
+
+
+\subsection nondet-transform Add Non-Deterministic Initialisation of Global Scoped Variables
+
+This transformation adds non-deterministic initialisation of global scoped
+variables including static variables. For details see
+`src/goto-instrument/nondet_static.h`. The initialisation code is added to the
+`CPROVER_initialize` function in the goto model. The implementation of this pass
+is called via the \ref nondet_static(goto_modelt &) function.
+
+<em>Predecessor pass is \ref update-transform.</em>
+
+
+\subsection unused-functions-transform Remove Unused Functions
+
```

Comment:
I think I responded to this observation previously. It is important to remove calls via function pointers before this pass because the remove unused functions pass will violate an INVARIANT if you do not.

Reply:
Sure, but one might argue then if you've had to respond to this before then kind of makes the point that readers could use a reminder. Maybe something like:
> **Note:** Removing unused functions before removing/lowering function pointers may result in false counter-examples.